<!doctype html><html lang='en'><head><title>Understanding errors | SuperCollider 3.11.2 Help</title>
<link rel='stylesheet' href='./../scdoc.css' type='text/css' />
<link rel='stylesheet' href='./../codemirror.css' type='text/css' />
<link rel='stylesheet' href='./../editor.css' type='text/css' />
<link rel='stylesheet' href='./../frontend.css' type='text/css' />
<link rel='stylesheet' href='./../custom.css' type='text/css' />
<meta name='viewport' content='width=device-width, initial-scale=1'>
<meta http-equiv='Content-Type' content='text/html; charset=UTF-8' />
<script src='./../lib/jquery.min.js'></script>
<script src='./../lib/codemirror-5.39.2.min.js' type='text/javascript'></script>
<script src='./../lib/codemirror-addon-simple-5.39.2.min.js' type='text/javascript'></script>
<script>
var helpRoot = './..';
var scdoc_title = 'Understanding errors';
var scdoc_sc_version = '3.11.2';
</script>
<script src='./../scdoc.js' type='text/javascript'></script>
<script src='./../docmap.js' type='text/javascript'></script>
<script src='qrc:///qtwebchannel/qwebchannel.js' type='text/javascript'></script>
</head>
<body onload='fixTOC()'>
<div id='toc'>
<div id='toctitle'>Understanding errors:</div>
<span class='toc_search'>Filter: <input id='toc_search'></span><ul class='toc'><li class='toc1'><a href='#Reading%20error%20dumps'>Reading error dumps</a></li>
<ul class='toc'></ul><li class='toc1'><a href='#Error%20objects%20and%20error%20handling'>Error objects and error handling</a></li>
<ul class='toc'></ul><li class='toc1'><a href='#Common%20primitive%20errors'>Common primitive errors</a></li>
<ul class='toc'><li class='toc2'><a href='#operation%20cannot%20be%20called%20from%20this%20Process.'>operation cannot be called from this Process.</a></li>
<ul class='toc'></ul><li class='toc2'><a href='#Attempted%20write%20to%20immutable%20object.'>Attempted write to immutable object.</a></li>
<ul class='toc'></ul><li class='toc2'><a href='#Index%20not%20an%20Integer.'>Index not an Integer.</a></li>
<ul class='toc'></ul><li class='toc2'><a href='#Index%20out%20of%20range.'>Index out of range.</a></li>
<ul class='toc'></ul></ul><li class='toc1'><a href='#A%20common%20warning'>A common warning</a></li>
<ul class='toc'></ul></ul></div><div id='menubar'></div>
<div class='contents'>
<div class='header'>
<div id='label'>
<span id='folder'>Guides</span>
 | <span id='categories'><a href='./../Browse.html#Language'>Language</a> | <a href='./../Browse.html#Debugging'>Debugging</a></span>
</div><h1>Understanding errors</h1>
<div id='summary'>a guide to understanding errors</div>
</div>
<div class='subheader'>
<div id='related'>See also: <a href="./../Guides/Debugging-tips.html">Debugging tips</a></div>
</div>
<h2><a class='anchor' name='Reading%20error%20dumps'>Reading error dumps</a></h2>

<p>When sc3 reports an error to the user, there are usually three parts:<ul>
<li>the error text<li>a dump of the receiver of the method that caused the error, and/or any arguments of the method call<li>a dump of the call stack to the point of the error</ul>

<p>For example:<textarea class='editor'>1.blech  // no class implements this method; therefore you'll get an error

// error text
ERROR: Message 'blech' not understood.

// receiver and args
RECEIVER:
   Integer 1
ARGS:
Instance of Array {    (02207560, gc=01, fmt=01, flg=11, set=00)
  indexed slots [0]
}

// call stack
CALL STACK:
    DoesNotUnderstandError-reportError
        arg this = &lt;instance of DoesNotUnderstandError&gt;
    Nil-handleError
        arg this = nil
        arg error = &lt;instance of DoesNotUnderstandError&gt;
    Object-throw
        arg this = &lt;instance of DoesNotUnderstandError&gt;
    Object-doesNotUnderstand
        arg this = 1
        arg selector = 'blech'
        arg args = [*0]
    &lt; closed FunctionDef &gt;  (no arguments or variables)
    Interpreter-interpretPrintCmdLine
        arg this = &lt;instance of Interpreter&gt;
        var res = nil
        var func = &lt;instance of Function&gt;
    Process-interpretPrintCmdLine
        arg this = &lt;instance of Main&gt;</textarea>

<p>Each of these parts provides valuable information about the cause of the error. Debugging is much easier if you understand what the error output means.<dl>
<dt>Error text<dd>A string describing the error. In this case, "Message 'xxx' not understood" means that you attempted to use the method xxx on a class that does not implement it.<dt>Receiver and arguments<dd>The method was applied to an integer (1), with no arguments (the size of the arguments array is 0).<dt>Call stack<dd>Order of execution in the call stack is in reverse: the top of the stack shows the most recent calls.</dl>

<p>Most call stacks for errors will show the same top three calls as shown here (calling the method reportError on an error class, calling handleError on Nil, and calling throw on the error object). You can ignore these three calls.
<p>Following is the meat: the error happened when an object was not understood. Continuing to read down, it happened inside a function definition. (Every time you highlight a block of code and press the enter key, the code is compiled into a function definition and executed. So, this function definition simply refers to the text submitted to the interpreter.) And, it all began with the instruction to interpret and print a command line.
<p>Here is a slightly more complex example, showing how you can use the variables listed for each call in the call stack to help locate the error.<textarea class='editor'>Routine({
    var a;
    a = 5;
    loop {
        var b;
        b = 20.rand;
        b.postln.ecky_ecky_phtang;   // "NI!!!!"
        a.wait;
    }
}).play;

ERROR: Message 'ecky_ecky_phtang' not understood.
RECEIVER:
   Integer 6
ARGS:
Instance of Array {    (02207560, gc=01, fmt=01, flg=11, set=00)
  indexed slots [0]
}
CALL STACK:
    DoesNotUnderstandError-reportError
        arg this = &lt;instance of DoesNotUnderstandError&gt;
    Nil-handleError
        arg this = nil
        arg error = &lt;instance of DoesNotUnderstandError&gt;
    Object-throw
        arg this = &lt;instance of DoesNotUnderstandError&gt;
    Object-doesNotUnderstand
        arg this = 6
        arg selector = 'ecky_ecky_phtang'
        arg args = [*0]
    &lt; FunctionDef in closed FunctionDef &gt;
        var b = 6
    Function-loop
        arg this = &lt;instance of Function&gt;
    &lt; FunctionDef in closed FunctionDef &gt;
        var a = 5
    Routine-prStart
        arg this = &lt;instance of Routine&gt;
        arg inval = 1542.075067</textarea>

<p>Reading from the bottom this time, to trace the flow in chronological order: this time, execution did not begin with the command line, but with a routine commencing within the scheduler (Routine({...}).play). Note that there are two calls identified as "FunctionDef in closed FunctionDef" and that they can be distinguished by the variables contained within. The earlier call (second from the bottom) defines the variable "a" while the other defines "b." To locate the error in the code, then, you should look for a function defining the variable "b" that is called within another function defining "a" inside a routine.
<p>What if the error occurred not inside a function definition that you wrote, but inside a method in the class library? There may be a bug in the method, or you may have thought the method took a certain kind of argument when in fact it expects something else.
<p>If you double click on the construction "ClassName-methodName" in the call stack, the whole thing is selected. Then you can press cmd-J to open the method definition and look at the source code.<h2><a class='anchor' name='Error%20objects%20and%20error%20handling'>Error objects and error handling</a></h2>

<p>sc3 implements error reporting using Error objects, which are instances of the class Error or one of its subclasses. Any code (whether in the class library or any user application) can throw an error any time as follows:<textarea class='editor'>Error("This is a basic error.").throw;</textarea>

<p>You can also catch exceptions that occur within functions by executing the function with <code>try</code> or <code>protect</code> instead of <code>value</code>.<dl>
<dt>try<dd>execute the first function. On an error, execute the second function and suppress the error. The second function can rethrow the error if desired, allowing you to decide which errors will be reported and which suppressed. In this example, we do not rethrow the error, so the error is swallowed and execution continues to the end.<textarea class='editor'>try { 1.blech } { |error| "oops".postln };
"next line".postln;

oops
next line</textarea>

<p><dt>protect<dd>executes the first function. On an error, execute the second function before reporting the error. This is useful when the steps before the protect make some changes that need to be undone if an error occurs. See <a href="./../Classes/Environment.html#use">Environment:use</a> for an example.<textarea class='editor'>protect { 1.blech } { |error| "oops".postln };
"next line".postln;

oops  // without protect, this would not be posted
ERROR: Message 'blech' not understood.
RECEIVER:
   Integer 1
ARGS:
Instance of Array {    (02207560, gc=01, fmt=01, flg=11, set=00)
  indexed slots [0]
}
CALL STACK:
    DoesNotUnderstandError-reportError
        arg this = &lt;instance of DoesNotUnderstandError&gt;</textarea>
</dl>

<p>Prior to August 2004, try and protect do not return the value of the function to the caller if there is no error.<textarea class='editor'>try { 1+1 }

a Function</textarea>

<p>More recent builds (since early August 2004) do return the function's value. Non-error objects can be thrown using the class Exception.<textarea class='editor'>try { 1+1 }
2

// can't add a Point to an integer - binary op failed error
// result of catch func is returned instead
try { 1+Point(0, 0) } { 2*5 }
10</textarea>
<h2><a class='anchor' name='Common%20primitive%20errors'>Common primitive errors</a></h2>
<h3><a class='anchor' name='operation%20cannot%20be%20called%20from%20this%20Process.'>operation cannot be called from this Process.</a></h3>

<p>This is usually the results of performing a GUI operation within a routine or scheduled function that is executing on some clock other than AppClock. AppClock is the only clock that can execute GUI manipulation because it is a lower priority thread. If the CPU is busy with audio synthesis or maintaining accurate scheduling for musical events, AppClock events will be delayed until the CPU is free enough.
<p>Solution: write your GUI updates as follows. defer schedules the function on AppClock.<textarea class='editor'>{ myGUIObject.value_(newValue) }.defer;</textarea>
<h3><a class='anchor' name='Attempted%20write%20to%20immutable%20object.'>Attempted write to immutable object.</a></h3>
<textarea class='editor'>#[0, 1, 2].put(1, 3)

ERROR: Primitive '_BasicPut' failed.
Attempted write to immutable object.</textarea>

<p><code>#[0, 1, 2]</code> is a literal array. Literal arrays cannot be manipulated--they can only be indexed. They cannot be changed internally.
<p>Solution: copy the array first.<textarea class='editor'>#[0, 1, 2].copy.put(1, 3)

[ 0, 3, 2 ]</textarea>
<h3><a class='anchor' name='Index%20not%20an%20Integer.'>Index not an Integer.</a></h3>
<textarea class='editor'>#[0, 1, 2].at(\1)

ERROR: Primitive '_BasicAt' failed.
Index not an Integer</textarea>

<p>Arrays can be indexed only with integers (or, in builds since August 2004, floats).
<p>Solution: use <code>.asInteger</code><textarea class='editor'>#[0, 1, 2].at(\1.asInteger)
1</textarea>

<p><div class='note'><span class='notelabel'>NOTE:</span> if the object cannot be converted into an integer, you'll get a "Does not understand" error!</div><h3><a class='anchor' name='Index%20out%20of%20range.'>Index out of range.</a></h3>
<textarea class='editor'>[0, 1, 2].put(5, 5)

ERROR: Primitive '_BasicPut' failed.
Index out of range.</textarea>

<p>Arrays have a finite size. If you try to put an object into an array slot but the slot does not exist because the array is too small, you'll get this error.
<p>Solution: extend the array.<textarea class='editor'>[0, 1, 2].extend(6).put(5, 5)

[ 0, 1, 2, nil, nil, 5 ]</textarea>

<p>Note that if the argument to extend() is smaller than the array, then the array will be truncated. If you're not sure, use max:<textarea class='editor'>i = rrand(5, 10);
a = [0, 1, 2];
a.extend(max(i+1, a.size)).put(i, 100);</textarea>

<p>Why i+1? An array with size 4 allows 0, 1, 2 and 3 as indexes (4 elements starting with 0).
<p>If it's a new array, use .newClear instead of .new.<textarea class='editor'>a = Array.new(4);
a.put(3, 1);
ERROR: Primitive '_BasicPut' failed.
Index out of range.

a = Array.newClear(4);
a.put(3, 1);
[ nil, nil, nil, 1 ]</textarea>
<h2><a class='anchor' name='A%20common%20warning'>A common warning</a></h2>
<textarea class='editor'>WARNING: FunctionDef contains variable declarations and so will not be inlined.</textarea>

<p>This warning can be safely ignored. Your code will still run, even if you get this warning.
<p>Inlining is a compiler optimization that takes the operations inside a function and places them in the main line of the containing function. For instance,<textarea class='editor'>// inlined
{ while { 0.9.coin } { 10.rand.postln }
}.def.dumpByteCodes;

BYTECODES: (16)
  0   40       PushLiteral Float 0.9   3FECCCCC CCCCCCCD  // { 0.9.coin }
  1   0D 2C    SendSpecialUnaryArithMsgX 'coin'
  3   F9 00 09 JumpIfFalsePushNil 9  (15)
  6   2C 0A    PushInt 10                              // { 10.rand.postln }
  8   0D 25    SendSpecialUnaryArithMsgX 'rand'
 10   C1 38    SendSpecialMsg 'postln'
 12   FD 00 0D JumpBak 13  (0)
 15   F2       BlockReturn
a FunctionDef in closed FunctionDef</textarea>

<p>This function contains two other functions. One is the condition for the while loop; the other is the while loop's action. The compiler renders this into a single code block, using jump instructions to handle the looping and exit.
<p>If, however, one of the functions defines a variable, then that function requires a separate execution frame. In this case, it's necessary for the compiler to push function definition objects onto the stack.<textarea class='editor'>// not inlined
{ while { 0.9.coin } {
    var a;    // variable here prevents optimization
    a = 10.rand;
    a.postln
  }
}.def.dumpByteCodes;

BYTECODES: (7)
  0   04 00    PushLiteralX instance of FunctionDef in closed FunctionDef
  2   04 01    PushLiteralX instance of FunctionDef in closed FunctionDef
  4   C2 0C    SendSpecialMsg 'while'
  6   F2       BlockReturn
a FunctionDef in closed FunctionDef</textarea>

<p>Inlined code will run faster, because pushing and using different execution frames is extra work for the virtual machine. If you're very concerned about speed, you can use this warning as an indicator that you might be able to optimize something in your code further.
<p>Sometimes, there's no way around un-optimized code. To wit,<textarea class='editor'>// inlined, optimized, but you'll get stuck notes
Routine({
  var synth;
  { synth = Synth("someSynth", [...args...]);
    thisThread.clock.sched(10, {
      synth.free;
    });
    2.wait;
  }.loop;
}).play;

// not inlined, but no stuck notes
Routine({
  { var synth;
    synth = Synth("someSynth", [...args...]);
    thisThread.clock.sched(10, {
      synth.free;
    });
    2.wait;
  }.loop;
}).play;</textarea>

<p>The first routine can be optimized because there is no variable declaration inside the loop. But, the synth variable changes on each iteration, meaning that by the time the first release happens, you don't have access anymore to the first note. Thus the first note will never terminate.
<p>In the second case, each note has its own synth variable, so the notes will be terminated as expected. You get a warning, but it's better because the results are correct.
<p>A solution to the above problem is to use a function with local variables.<textarea class='editor'>(
Routine({
    var func;
    func = {
        var synth; // this variable is local to the function
        synth = Synth("default");
        [\play, synth].postln;
        thisThread.clock.sched(4.5, {
            synth.free;
            [\free, synth].postln;
        });
    };
    { func.value; 1.wait; }.loop
}).play;
)</textarea>
<div class='doclink'>helpfile source: <a href='file:///usr/share/SuperCollider/HelpSource/Guides/Understanding-Errors.schelp'>/usr/share/SuperCollider/HelpSource/Guides/Understanding-Errors.schelp</a><br>link::Guides/Understanding-Errors::<br></div></div><script src='./../editor.js' type='text/javascript'></script>
</body></html>